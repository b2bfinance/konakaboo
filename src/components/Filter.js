import React, { Component } from 'react';
import { connect } from 'react-redux';
import { queries } from '../utils/media';
import { getChosenForGroup } from '../utils/filter';
import { setChosenFiltersForGroup } from '../actions/filter';
import styled, { css } from 'styled-components';
import Col from './Col';
import { FormControl } from 'material-ui/Form';
import Input, { InputLabel } from 'material-ui/Input';
import Select from 'material-ui/Select';
import { MenuItem } from 'material-ui/Menu';

export const FilterWrapper = Col.extend`
  border-radius: 0.2rem;
  margin-bottom: 1rem;
  position: relative;

  ${queries.tablet`
    padding-right: 1rem;

    &:last-child {
      padding-right: 0;
    }
  `};
`;

export const FilterFormControl = styled(FormControl)`
  width: 100%;
`;

export const FilterSelect = styled(Select)`
  &:hover {
    &:before {
      background-color: ${props => props.theme.filterHeaderBorder} !important;
    }
  }

  &:after {
    background-color: ${props => props.theme.filterHeaderBorder} !important;
  }
`;

export const FilterMenuItem = styled(MenuItem)`
  ${props =>
    props.selected &&
    css`
      background-color: ${props.theme.filterChosenBackground} !important;
      color: ${props.theme.filterChosenColor} !important;
    `};
`;

export class Filter extends Component {
  render() {
    const { filters, chosen, handleSelectChange } = this.props;

    return filters.map(({ title, multiChoice, choices, key }, i) => (
      <FilterWrapper key={key} phone="100" tablet="33.333" desktop="20">
        <FilterFormControl>
          <InputLabel htmlFor={key}>{title}</InputLabel>
          <FilterSelect
            multiple={multiChoice}
            value={getChosenForGroup(chosen, i)}
            onChange={e => handleSelectChange(i, e.target.value)}
            input={<Input id={key} />}
          >
            {choices.map(({ label, value }) => (
              <FilterMenuItem key={value} value={value}>
                {label}
              </FilterMenuItem>
            ))}
          </FilterSelect>
        </FilterFormControl>
      </FilterWrapper>
    ));
  }
}

export const mapDispatchToProps = {
  handleSelectChange: setChosenFiltersForGroup
};

export const mapStateToProps = ({ filters }) => ({
  chosen: filters.chosen,
  filters: filters.available
});

export default connect(mapStateToProps, mapDispatchToProps)(Filter);
