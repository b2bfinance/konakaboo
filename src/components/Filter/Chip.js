import styled, { css } from 'styled-components';
import Chip from '@material-ui/core/Chip';

export { Chip as BaseChip };

export const ChipHolder = styled.div`
  margin-bottom: 1.5rem;
  margin-right: 0.5rem;
  position: relative;
`;

export default styled(Chip)`
  ${props =>
    props.selection.length > 0 &&
    css`
      background-color: ${props.theme.filterChipChosenBackground} !important;
      color: ${props.theme.filterChipChosenColor} !important;
    `};
`;
