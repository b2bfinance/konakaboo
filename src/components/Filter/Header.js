import styled from 'styled-components';

export default styled.div`
  align-items: center;
  background-color: ${props => props.theme.filterHeaderBackground};
  color: ${props => props.theme.filterHeaderColor};
  display: flex;
  font-weight: ${props => props.theme.mainBoldFontWeight};
  justify-content: space-between;
  min-width: 15rem;
  padding: 0.45rem 0.5rem 0.45rem 0.75rem;

  svg {
    fill: ${props => props.theme.filterHeaderColor};
  }
`;
