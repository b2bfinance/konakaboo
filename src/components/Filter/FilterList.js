import React, { Component } from 'react';
import { connect } from 'react-redux';
import Filter from './Filter';

class FilterList extends Component {
  render() {
    const { groups } = this.props;

    return (
      groups.map((filter, i) => (
        <Filter key={i} filter={filter} groupIndex={i}/>
      ))
    );
  }
}

const mapStateToProps = ({filters}) => ({
  groups: filters.groups,
});

export default connect(mapStateToProps)(FilterList);
