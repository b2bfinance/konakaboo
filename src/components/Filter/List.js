import Grid from '@material-ui/core/Grid';
import ArrowDropDown from '@material-ui/icons/ArrowDropDown';
import Cancel from '@material-ui/icons/Cancel';
import React, { useState } from 'react';
import styled from 'styled-components';
import {
  resetFilters,
  resetGroupFilters,
  setFilters
} from '../../actions/filter';
import { generateChipLabel } from '../../utils/filter';
import Chip, { BaseChip, ChipHolder } from './Chip';
import Filter from './Filter';

export const Wrapper = styled(Grid)`
  justify-content: flex-end;
  margin-bottom: 32px;
`;

export default ({ filters, dispatch }) => {
  const [activeGroup, setActiveGroup] = useState();

  const handleDeleteChosen = (group, key) => {
    if (filters.chosen[group].length) {
      if (activeGroup === key) {
        setActiveGroup();
      }

      dispatch(resetGroupFilters(group));
      return;
    }

    setActiveGroup(key);
  };

  return (
    <Wrapper container>
      {filters.available.map((filter, i) => (
        <ChipHolder key={filter.key}>
          <Chip
            selection={filters.chosen[i]}
            label={generateChipLabel(
              filter.title,
              filter.multiChoice,
              filters.chosen[i],
              filter.choices
            )}
            deleteIcon={
              filters.chosen[i].length ? <Cancel /> : <ArrowDropDown />
            }
            onClick={() => setActiveGroup(filter.key)}
            onDelete={() => handleDeleteChosen(i, filter.key)}
            onBlur={() => setActiveGroup()}
          />
          <Filter
            visible={activeGroup === filter.key}
            group={i}
            multi={filter.multiChoice}
            title={filter.title}
            choices={filter.choices}
            chosen={filters.chosen[i]}
            handleSetChosenForGroup={(group, chosen) =>
              dispatch(setFilters(group, chosen))
            }
            handleClose={() => setActiveGroup()}
          />
        </ChipHolder>
      ))}
      <BaseChip
        label="Reset Filters"
        onClick={() => dispatch(resetFilters())}
      />
    </Wrapper>
  );
};
