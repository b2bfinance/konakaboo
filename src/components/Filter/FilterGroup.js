import React from 'react';
import { connect } from 'react-redux';
import { removeGroupFilters } from '../../actions/filter';
import { queries } from '../../utils/media';
import styled from 'styled-components';
import Button from '../Button';
import Clear from 'material-ui-icons/Clear';
import FilterList from './FilterList';
import Col from '../Col';

const FilterWrapper = Col.extend`
  border-radius: .2rem;
  margin-bottom: 1rem;

  ${queries.tablet`
    padding-right: 1rem;

    &:last-child {
      padding-right: 0;
    }
  `}
`

const FilterHeader = styled.div`
  align-items: center;
  background-color: ${props => props.theme.filterHeaderBackground};
  color: ${props => props.theme.filterHeaderColor};
  display: flex;
  justify-content: space-between;
  padding: .45rem .65rem;
`

const FilterBody = styled.div`
  background-color: ${props => props.theme.filterBodyColor};
`

const FilterFooter = FilterBody.extend`
  cursor: pointer;
  border-top: 1px solid ${props => props.theme.filterFooterBorderColor};
  padding: .35rem;
  text-decoration: underline;
  text-align: center;
`

const FilterGroup = ({
  filter,
  group,
  onClearFilters,
}) => (
  <FilterWrapper phone="100" tablet="50" desktop="25">
    <FilterHeader>
      <span>{filter.title}</span>
    </FilterHeader>
    <FilterBody>
      <FilterList
        multi={filter.multiChoice}
        choices={filter.choices}
        group={group}
       />
    </FilterBody>
    <FilterFooter onClick={() => onClearFilters(group)}>
      Clear Filters
    </FilterFooter>
  </FilterWrapper>
);

const mapDispatchToProps = {
  onClearFilters: removeGroupFilters,
};

export default connect(() => ({}), mapDispatchToProps)(FilterGroup);
