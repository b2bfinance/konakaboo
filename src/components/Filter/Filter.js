import React from 'react';
import styled, { css } from 'styled-components';
import { queries } from '../../utils/media';
import { getChosenWithModification } from '../../utils/filter';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import IconButton from '@material-ui/core/IconButton';
import Checkbox from '@material-ui/core/Checkbox';
import Radio from '@material-ui/core/Radio';
import Close from '@material-ui/icons/Close';
import Header from './Header';
import { Typography } from '@material-ui/core';

export const Wrapper = styled.div`
  background-color: #fff;
  border-radius: 0.2rem;
  box-shadow: 0 15px 12px rgba(0, 0, 0, 0.22), 0 0 38px rgba(0, 0, 0, 0.3);
  margin-top: 0.3rem;
  min-width: 100%;
  opacity: 0;
  right: 0;
  transform-origin: right top 0;
  transform: scale3d(0.8, 0.8, 1);
  transition: opacity 250ms cubic-bezier(0.8, 0, 0.6, 1) 0ms,
    transform 250ms cubic-bezier(0.8, 0, 0.6, 1) 0ms, visibility 250ms;
  visibility: hidden;
  z-index: 150;

  ${props =>
    props.visible &&
    css`
      opacity: 1;
      transform: scale3d(1, 1, 1);
      visibility: visible;
    `};

  ${queries.phone`
    bottom: 0;
    position: fixed;
  `};

  ${queries.tablet`
    bottom: auto;
    position: absolute;
  `};

  ul {
    max-height: 25.5rem;
    overflow: auto;
  }
`;

export const Heading = styled(Typography)`
  color: ${props => props.theme.filterHeaderColor} !important;
`;

export function handleMouseDown(e) {
  e.preventDefault();
}

export default ({
  visible,
  group,
  multi,
  title,
  choices,
  chosen,
  handleSetChosenForGroup,
  handleClose
}) => (
  <Wrapper visible={visible} onMouseDown={handleMouseDown}>
    <Header>
      <Heading variant="subheading">{title}</Heading>
      <IconButton onClick={handleClose} aria-label="Close Filter Choices">
        <Close />
      </IconButton>
    </Header>
    <List>
      {choices.map(choice => (
        <ListItem
          key={choice.value}
          button
          onClick={() =>
            handleSetChosenForGroup(
              group,
              getChosenWithModification(chosen, multi, choice.value)
            )
          }
        >
          {multi ? (
            <Checkbox
              checked={chosen.indexOf(choice.value) !== -1}
              tabIndex={-1}
              disableRipple
            />
          ) : (
            <Radio
              checked={chosen === choice.value}
              tabIndex={-1}
              disableRipple
            />
          )}
          <ListItemText primary={choice.label} />
        </ListItem>
      ))}
    </List>
  </Wrapper>
);
