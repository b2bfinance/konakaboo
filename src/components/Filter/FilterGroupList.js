import React from 'react';
import { connect } from 'react-redux';
import FilterGroup from './FilterGroup';

const FilterGroupList = ({filters}) => filters.available.map((filter, i) => (
  <FilterGroup key={i} filter={filter} group={i} />
));

export default connect(({filters}) => ({filters}))(FilterGroupList);
