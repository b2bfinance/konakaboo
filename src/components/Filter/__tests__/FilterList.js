import React from 'react';
import { mount, shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import FilterList, {
  FilterRow,
  ChipContainer,
  StyledChip,
  generateChipLabel,
  mapDispatchToProps,
  mapStateToProps
} from '../FilterList';
import { Filter } from '../Filter';
import Chip from '@material-ui/core/Chip';
import Cancel from '@material-ui/icons/Cancel';

test('<FilterRow /> renders correctly', () => {
  expect(shallow(<FilterRow />)).toMatchSnapshot();
});

test('<ChipContainer /> renders correctly', () => {
  expect(shallow(<ChipContainer />)).toMatchSnapshot();
});

describe('<StyledChip />', () => {
  test('with no selection renders correctly', () => {
    expect(shallow(<StyledChip selection={[]} />)).toMatchSnapshot();
  });

  test('with selection renders correctly', () => {
    expect(shallow(<StyledChip selection={['TEST']} />)).toMatchSnapshot();
  });
});

describe('generateChipLabel', () => {
  const choices = [
    { label: 'Chosen Label', value: 'chosen_1' },
    { label: 'Chosen Label', value: 'chosen_2' },
    { label: 'Chosen Label', value: 'chosen_3' }
  ];

  test('creates a none multi label with no chosen', () => {
    expect(generateChipLabel('label', false, '', choices)).toBe('label');
  });

  test('creates a multi label with no chosen', () => {
    expect(generateChipLabel('label', true, [], choices)).toBe('label');
  });

  test('creates a none multi label with chosen', () => {
    expect(generateChipLabel('label', false, 'chosen_1', choices)).toBe(
      'Chosen Label'
    );
  });

  test('creates a multi label with chosen', () => {
    expect(
      generateChipLabel('label', true, ['chosen_1', 'chosen_2'], choices)
    ).toBe('Chosen Label +1');
  });
});

describe('<FilterList />', () => {
  let store;
  let wrapper;

  beforeEach(() => {
    const mockStore = configureMockStore([thunk]);
    const mockState = stubData.filters.withSingleAndMultiChoiceChosen;

    store = mockStore({ config: { provider: '' }, filters: mockState });

    wrapper = mount(
      <Provider store={store}>
        <FilterList />
      </Provider>
    );
  });

  test('has the correct chip labels', () => {
    const chips = wrapper.find(StyledChip);
    expect(chips.first().text()).toBe('TEST_CHOICE_1_LABEL +1');
    expect(chips.last().text()).toBe('TEST_CHOICE_1_LABEL');
  });

  test('will toggle filter visiblilty when clicking associated chip', () => {
    const chip = wrapper.find(StyledChip).first();

    chip.simulate('click');

    expect(
      wrapper
        .find(Filter)
        .first()
        .prop('visible')
    ).toBe(true);

    chip.simulate('click');

    expect(
      wrapper
        .find(Filter)
        .first()
        .prop('visible')
    ).toBe(false);
  });

  test('will dispatch the correct actions when deleting', () => {
    wrapper
      .find(Cancel)
      .first()
      .simulate('click');

    expect(store.getActions()).toEqual([
      { type: 'RESET_GROUP_FILTERS', group: 0, payload: {} },
      { type: 'PRODUCTS_REQUEST' }
    ]);
  });

  test('will close the filter when clicking the associated delete element', () => {
    wrapper
      .find(Chip)
      .first()
      .simulate('click');

    wrapper
      .find(Cancel)
      .first()
      .simulate('click');

    expect(
      wrapper
        .find(Filter)
        .first()
        .prop('visible')
    ).toBe(false);
  });

  test('will reset filters when reset chip is clicked', () => {
    wrapper
      .find(Chip)
      .last()
      .simulate('click');

    expect(store.getActions()).toEqual([
      { payload: {}, type: 'RESET_FILTERS' },
      { type: 'PRODUCTS_REQUEST' }
    ]);
  });
});

test('mapDispatchToProps maps the dispatch correctly', () => {
  expect(Object.keys(mapDispatchToProps)).toEqual([
    'handleResetFiltersForGroup',
    'handleResetFilters'
  ]);
});

test('mapStateToProps maps the state correctly', () => {
  const stubState = stubData.filters.withSingleAndMultiChoiceChosen;

  expect(mapStateToProps({ filters: stubState })).toEqual({
    chosen: stubState.chosen,
    filters: stubState.available
  });
});
