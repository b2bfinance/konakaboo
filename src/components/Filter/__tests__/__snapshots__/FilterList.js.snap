// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChipContainer /> renders correctly 1`] = `
.c0 {
  margin-bottom: 1.5rem;
  margin-right: 0.5rem;
  position: relative;
}

<div
  className="c0"
/>
`;

exports[`<StyledChip /> with no selection renders correctly 1`] = `
<withStyles(Chip)
  className=""
  selection={Array []}
/>
`;

exports[`<StyledChip /> with selection renders correctly 1`] = `
.c0 {
  background-color: !important;
  color: !important;
}

<withStyles(Chip)
  className="c0"
  selection={
    Array [
      "TEST",
    ]
  }
/>
`;
